File: src/main/java/com/indiepost/dto/post/PostQuery.java
package com.indiepost.dto.post;

import java.time.Instant;
import java.time.LocalDateTime;

import static com.indiepost.enums.Types.PostStatus;
import static com.indiepost.utils.DateUtil.instantToLocalDateTime;

/**
 * Created by jake on 17. 1. 10.
 */
public class PostQuery {

    private String category;

    private String displayName;

    private String text;

    private PostStatus status;

    private Boolean featured;

    private Boolean picked;

    private Boolean splash;

    private LocalDateTime modifiedAfter;

    private LocalDateTime modifiedBefore;

    private LocalDateTime publishedAfter;

    private LocalDateTime publishedBefore;

    private PostQuery() {
    }

    public String getCategory() {
        return category;
    }

    public String getDisplayName() {
        return displayName;
    }

    public String getText() {
        return text;
    }

    public PostStatus getStatus() {
        return status;
    }

    public Boolean getFeatured() {
        return featured;
    }

    public Boolean getPicked() {
        return picked;
    }

    public Boolean getSplash() {
        return splash;
    }

    public LocalDateTime getModifiedAfter() {
        return modifiedAfter;
    }

    public LocalDateTime getModifiedBefore() {
        return modifiedBefore;
    }

    public LocalDateTime getPublishedAfter() {
        return publishedAfter;
    }

    public LocalDateTime getPublishedBefore() {
        return publishedBefore;
    }

    public static class Builder {
        private PostStatus status;

        private String category;

        private String displayName;

        private String text;

        private Boolean featured;

        private Boolean picked;

        private Boolean splash;

        private Instant modifiedAfter;

        private Instant modifiedBefore;

        private Instant publishedAfter;

        private Instant publishedBefore;

        public Builder(PostStatus status) {
            this.status = status;
        }

        public Builder(String status) {
            this.status = PostStatus.valueOf(status);
        }

        public Builder category(String category) {
            this.category = category;
            return this;
        }

        public Builder displayName(String displayName) {
            this.displayName = displayName;
            return this;
        }

        public Builder searchText(String text) {
            this.text = text;
            return this;
        }

        public Builder featured(boolean isFeatured) {
            this.featured = isFeatured;
            return this;
        }

        public Builder picked(boolean isPicked) {
            this.picked = isPicked;
            return this;
        }

        public Builder splash(boolean isSplash) {
            this.splash = isSplash;
            return this;
        }

        public Builder publishedAfter(Instant publishedAfter) {
            this.publishedAfter = publishedAfter;
            return this;
        }

        public Builder publishedBefore(Instant publishedBefore) {
            this.publishedBefore = publishedBefore;
            return this;
        }

        public Builder modifiedAfter(Instant modifiedAfter) {
            this.modifiedAfter = modifiedAfter;
            return this;
        }

        public Builder modifiedBefore(Instant modifiedBefore) {
            this.modifiedBefore = modifiedBefore;
            return this;
        }

        public PostQuery build() {
            PostQuery query = new PostQuery();

            // Required
            query.status = status;

            if (category != null) {
                query.category = category;
            }
            if (displayName != null) {
                query.displayName = displayName;
            }
            if (text != null) {
                query.text = text;
            }
            if (featured != null) {
                query.featured = featured;
            }
            if (picked != null) {
                query.picked = picked;
            }
            if (splash != null) {
                query.splash = splash;
            }
            if (publishedAfter != null) {
                query.publishedAfter = instantToLocalDateTime(publishedAfter);
            }
            if (publishedBefore != null) {
                query.publishedBefore = instantToLocalDateTime(publishedBefore);
            }
            if (modifiedAfter != null) {
                query.modifiedAfter = instantToLocalDateTime(modifiedAfter);
            }
            if (modifiedBefore != null) {
                query.modifiedBefore = instantToLocalDateTime(modifiedBefore);
            }
            return query;
        }
    }
}

File: src/main/kotlin/com/indiepost/dto/post/PostQuery.kt
package com.indiepost.dto.post

import com.indiepost.enums.Types.PostStatus
import com.indiepost.utils.DateUtil.instantToLocalDateTime
import java.time.Instant
import java.time.LocalDateTime

/**
 * Created by jake on 17. 1. 10.
 */
class PostQuery private constructor() {

    var category: String? = null
        private set

    var displayName: String? = null
        private set

    var text: String? = null
        private set

    var status: PostStatus? = null
        private set

    var featured: Boolean? = null
        private set

    var picked: Boolean? = null
        private set

    var splash: Boolean? = null
        private set

    var modifiedAfter: LocalDateTime? = null
        private set

    var modifiedBefore: LocalDateTime? = null
        private set

    var publishedAfter: LocalDateTime? = null
        private set

    var publishedBefore: LocalDateTime? = null
        private set

    class Builder {
        private var status: PostStatus? = null

        private var category: String? = null

        private var displayName: String? = null

        private var text: String? = null

        private var featured: Boolean? = null

        private var picked: Boolean? = null

        private var splash: Boolean? = null

        private var modifiedAfter: Instant? = null

        private var modifiedBefore: Instant? = null

        private var publishedAfter: Instant? = null

        private var publishedBefore: Instant? = null

        constructor(status: PostStatus) {
            this.status = status
        }

        constructor(status: String) {
            this.status = PostStatus.valueOf(status)
        }

        fun category(category: String): Builder {
            this.category = category
            return this
        }

        fun displayName(displayName: String): Builder {
            this.displayName = displayName
            return this
        }

        fun searchText(text: String): Builder {
            this.text = text
            return this
        }

        fun featured(isFeatured: Boolean): Builder {
            this.featured = isFeatured
            return this
        }

        fun picked(isPicked: Boolean): Builder {
            this.picked = isPicked
            return this
        }

        fun splash(isSplash: Boolean): Builder {
            this.splash = isSplash
            return this
        }

        fun publishedAfter(publishedAfter: Instant): Builder {
            this.publishedAfter = publishedAfter
            return this
        }

        fun publishedBefore(publishedBefore: Instant): Builder {
            this.publishedBefore = publishedBefore
            return this
        }

        fun modifiedAfter(modifiedAfter: Instant): Builder {
            this.modifiedAfter = modifiedAfter
            return this
        }

        fun modifiedBefore(modifiedBefore: Instant): Builder {
            this.modifiedBefore = modifiedBefore
            return this
        }

        fun build(): PostQuery {
            val query = PostQuery()

            // Required
            query.status = status

            if (category != null) {
                query.category = category
            }
            if (displayName != null) {
                query.displayName = displayName
            }
            if (text != null) {
                query.text = text
            }
            if (featured != null) {
                query.featured = featured
            }
            if (picked != null) {
                query.picked = picked
            }
            if (splash != null) {
                query.splash = splash
            }
            publishedAfter?.let {
                query.publishedAfter = instantToLocalDateTime(it)
            }
            publishedBefore?.let {
                query.publishedBefore = instantToLocalDateTime(it)
            }
            modifiedBefore?.let {
                query.modifiedBefore = instantToLocalDateTime(it)
            }
            modifiedAfter?.let {
                query.modifiedAfter = instantToLocalDateTime(it)
            }
            return query
        }
    }
}
